iwyu_output.cc

```diff
@@ -2298,4 +2306,19 @@ size_t IwyuFileInfo::CalculateAndReportIwyuViolations() {
   return num_edits;
 }
 
+bool IsDirectlyIncluded(const NamedDecl* decl, const IwyuFileInfo& includer) {
```

Comment:
This feels like it belongs more with `IwyuPreprocessorInfo` or `IncludePicker` than the output module (it seems suspicious for `iwyu.cc` to ask `output` about core decisions). But this is really just a feeling, there may be good reasons to keep it in iwyu_output that I'm not seeing.

Reply:
I agree that it is strange place for it. The reason to insert it here was in that `IsDirectlyIncluded` takes `IwyuFileInfo` and could be considered as part of its interface and hence be placed next to it.

I've moved `GetMappedPublicHeaders` to `iwyu_include_picker` and `IsDirectlyIncluded` to `iwyu_preprocessor` (placing `IsDirectlyIncluded` into `iwyu_include_picker` would require additional dependency for the latter from `iwyu_output.h`). Still not sure that it is the best decision.

Reply:
Thanks for rationalizing, now I see it more clearly. I think the best would be:

* Make `GetMappedPublicHeaders` a member function of `IncludePicker` (would obviate the need to reach out to `GlobalIncludePicker` for itself :))
* Move `IsDirectlyIncluded` back into `iwyu_output`

I think that would make most sense.

Reply:
Authorities on C++ claim that [if a function can be nonmember nonfriend then it should be](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c4-make-a-function-a-member-only-if-it-needs-direct-access-to-the-representation-of-a-class). There is also [Herb Sutter's article](http://www.gotw.ca/gotw/084.htm) about it.

Reply:
Yeah, that makes sense for value types in general. But the dependency on `GlobalIncludePicker` in `GetMappedPublicHeaders` makes it almost absurd. We want clients to be able to use whatever `IncludePicker` they have available (which in most cases will be `GlobalIncludePicker()`, but that could change). So either we'd pass in `IncludePicker&` to `GetMappedPublicHeaders`, or we make it a member. The most obvious, IMO, is a member.

Reply:
If `IsDirectlyIncluded` will be returned into `iwyu_output`, then `GetMappedPublicHeaders()` can be made `static` function of `iwyu_output.cc` again, in principle...

Reply:
Yeah, but I think it makes more sense as a member of `IncludePicker`, since it owns all the mappings